### PR TITLE
sql: improve the ROUND and EXP functions.

### LINKED ISSUE #8633: sql: decimal round can DOS the server
```
SELECT round(-0.32096519520627553, -2438602134409251682)
```

The above query will peg one CPU core at 100% for many minutes until it completes (I have not run it long enough to see how long that is). This is because the implementation of decimal rounding does some quotient remainder calculations that take a while or something.

In any case it's not obvious to me how to handle this, but something should be done since this makes it trivial to shut down a server.

Found with RSG.

### PR DESCRIPTION
**Pull request by knz**
Prior to this patch the computation would take too long or too much
memory if EXP was given an argument too large or if the 2nd ROUND
argument was negative with a large absolute value.

This patch addresses this issue by limiting the range of acceptable
arguments (1024 for EXP, anything that causes more than 2000 extra "0"
digits to be appended on the right for ROUND).

It improves both functions as follows:
- a negative ROUND scale is now supported as in PostgreSQL and Oracle;
  this rounds to positions on the left of the decimal separator.
- the ROUND function for floats is now performed on floats directly,
  instead of using Printf/ParseFloat as previously (faster).
- the ROUND function for decimals is now faster if the 2nd argument is
  negative and its absolute value is larger than the number of digits
  in the 1st argument on the left of its decimal separator.
- ROUND for both floats and decimal now use half to even rounding,
  also called unbiased rounding or bankers' rounding.

Fixes #8633.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8822)

### CODE REVIEW

File: sql/parser/builtins.go [hunk 1]
```diff
@@ -1405,32 +1401,79 @@ func overlay(s, to string, pos, size int) (Datum, error) {
...
+
+	return NewDFloat(DFloat(v / pow)), nil
+}
+
+const (
+	scaleRatio = math.Ln2 / math.Ln10
+	// Compute the size of a Word in bits.
+	_m       = ^big.Word(0)
+	_logS    = _m>>8&1 + _m>>16&1 + _m>>32&1
+	wordSize = 8 * (1 << _logS)
```
Comment (knz) [inline sql/parser/builtins.go]: Any suggestion as to how to implement this without having `go vet` bark on `_m >> 32` like it currently does?

**Comment by mjibson** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/builtins.go, line 1436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ18qq-rqz1xgqhDRva-r1-1436:-KQ1uZHbmbhJnFysduii:b-n7zbyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1436)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Any suggestion as to how to implement this without having `go vet` bark on `_m >> 32` like it currently does?
</details>
We're lucky: I wrote that vet test. `_m = int64(^big.Word(0))` will force _m to be an int64 instead of int. int can be 32-bits on some systems, which is why vet is complaining. Crazy that it's the same code in math/big. I guess they ignore some vet warnings?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-:-KQ1zoD4oZjd5N3rp5Ue:bnfp4nl)*

**Comment by nvanbenschoten** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1xph6YfCcaEOHRicx:-KQ1xph6YfCcaEOHRicy:bow85z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L52)):*
> ```Go
> 	errEmptyInputString = errors.New("the input string must not be empty")
> 	errAbsOfMinInt64    = errors.New("abs of min integer value (-9223372036854775808) not defined")
> 	errRoundTooLow      = errors.New("rounding would extend value by more than 2000 decimals")
> ```

`s/2000 decimals/2000 digits`?

---

*[sql/parser/builtins.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1xwuEAYzjRkO4Z-Gw:-KQ1xwuEAYzjRkO4Z-Gx:b-qk2vpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L53)):*
> ```Go
> 	errAbsOfMinInt64    = errors.New("abs of min integer value (-9223372036854775808) not defined")
> 	errRoundTooLow      = errors.New("rounding would extend value by more than 2000 decimals")
> 	errArgTooBig        = errors.New("argument value is too large")
> ```

I wonder if this should say the limit to be more helpful.

---

*[sql/parser/builtins.go, line 1406 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yGMH18xlY-grW3xL:-KQ1yGMH18xlY-grW3xM:bfryc9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1406)):*
> ```Go
> 	pow := math.Pow(10, float64(n))
> 
> 	if pow == 0 {
> ```

my personal preference would be a switch statement to handle all edge cases here, but it's up to you

---

*[sql/parser/builtins.go, line 1419 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yRxc2RtpvSBc7g4M:-KQ1yRxc2RtpvSBc7g4N:bim9vwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1419)):*
> ```Go
> 	v, frac := math.Modf(x * pow)
> 	if x > 0.0 {
> 		if frac > 0.5 || (frac == 0.5 && uint64(v)%2 != 0) {
> ```

You might want to add a comment about the `(frac == 0.5 && uint64(v)%2 != 0)` here and below. Something about SQL rounding and how evens are rounded up, etc. I'm not sure if there's a specific name for this, but I've heard it called "unbiased rounding".

---

*[sql/parser/builtins.go, line 1436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ18qq-rqz1xgqhDRva-r1-1436:-KQ1zTDpAqSA56jhnyoF:b-ei392m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1436)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Any suggestion as to how to implement this without having `go vet` bark on `_m >> 32` like it currently does?
</details>
We have a lot of this in `util/encoding/decimal.go`. You can just use `int(unsafe.Sizeof(big.Word(0)))`, which has the same effect.

---

*[sql/parser/builtins.go, line 1444 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1zzn5al84VorX48vH:-KQ1zzn6pRZfDiK6Copb:bq2bwiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1444)):*
> ```Go
> 	if n > curScale+2000 {
> 		// If we let the decimal value grow too many decimals, the server
> 		// could explode (#8633).
> ```



---

*[sql/parser/builtins.go, line 1453 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ2-_LdLg1W6LQ87lOc:-KQ2-_LdLg1W6LQ87lOd:b-4t2j59) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1453)):*
> ```Go
> 	upperDigitsLeft := float64(curScale) - float64(upperCurDigits)*scaleRatio
> 	if n >= int64(upperDigitsLeft)-1 {
> 		// This is an optimization. When the rounding scale is definitely
> ```

I would change this wording or reverse the inequality. You're talking about `definitely larger` here, so it made me think you would enter this block for the optimization and that `n` was the rounding scale. Instead, we avoid this block for the optimization.

---

*[sql/parser/builtins.go, line 1464 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ20AD8wBqdHkr0FSB0:-KQ20AD8wBqdHkr0FSB1:b-frgibw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1464)):*
> ```Go
> 	// The computation of Exp is separated in the decimal module by
> 	// computing the exponents on the left and right of the decimal
> 	// separator.  The computation on the right is bounded by
> ```

`s/  / /` (two spaces -> one)

---

*[sql/testdata/builtin_function, line 686 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ20j34vCLZJN6jBBBw:-KQ20j34vCLZJN6jBBBx:b-yey2sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/testdata/builtin_function#L686)):*
> ```
> 
> query error rounding would extend value by more than 2000 decimals
> SELECT round(1::decimal, 3000)
> ```

We should add a test for the `errArgTooBig` case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 1419 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yRxc2RtpvSBc7g4M:-KQ21jjsdO_2htyolrIS:bxwqy4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1419)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> You might want to add a comment about the `(frac == 0.5 && uint64(v)%2 != 0)` here and below. Something about SQL rounding and how evens are rounded up, etc. I'm not sure if there's a specific name for this, but I've heard it called "unbiased rounding".
</details>
Also called bankers' rounding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1xph6YfCcaEOHRicx:-KQ2hAoDZtOX41iPPoEW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/2000 decimals/2000 digits`?
</details>
Done.

---

*[sql/parser/builtins.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1xwuEAYzjRkO4Z-Gw:-KQ2hBkSlt2HzdmDLhGV:bkpzrma) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L53)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I wonder if this should say the limit to be more helpful.
</details>
I'm not yet comfortable with the specific limit I chose so I'd rather not advertise it for now.

---

*[sql/parser/builtins.go, line 1406 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yGMH18xlY-grW3xL:-KQ2hMhnbzU60JE-Hhoz:b-is0usg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1406)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> my personal preference would be a switch statement to handle all edge cases here, but it's up to you
</details>
I'm not comfortable using switch here because switch suggests the conditions are mutually exclusive and thus that the ordering of tests is indeterminate. Here the conditions have to be evaluated in sequence.

---

*[sql/parser/builtins.go, line 1419 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yRxc2RtpvSBc7g4M:-KQ2ictng4ReRWbaIsIX:beibg4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1419)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Also called bankers' rounding.
</details>
Done. Also changed roundDecimal to use the same for consistency.

---

*[sql/parser/builtins.go, line 1436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ18qq-rqz1xgqhDRva-r1-1436:-KQ2j6JPC8v5_WMU5t4o:bajavzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1436)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> We have a lot of this in `util/encoding/decimal.go`. You can just use `int(unsafe.Sizeof(big.Word(0)))`, which has the same effect.
</details>
Thanks for the tip! I'm using the Sizeof solution now.

---

*[sql/parser/builtins.go, line 1444 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1zzn5al84VorX48vH:-KQ2jTwbrS5wrRc5-ESm:b7bl4kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1444)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> 
</details>
:fire: :fire: :fire: 

---

*[sql/parser/builtins.go, line 1453 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ2-_LdLg1W6LQ87lOc:-KQ2jplMzpRtryTkd4i4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1453)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I would change this wording or reverse the inequality. You're talking about `definitely larger` here, so it made me think you would enter this block for the optimization and that `n` was the rounding scale. Instead, we avoid this block for the optimization.
</details>
Done.

---

*[sql/parser/builtins.go, line 1464 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ20AD8wBqdHkr0FSB0:-KQ2jvOR7PwSZ7u_QKRi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1464)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/  / /` (two spaces -> one)
</details>
Done.

---

*[sql/testdata/builtin_function, line 686 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ20j34vCLZJN6jBBBw:-KQ2jwg0pLV-src_wBZN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/testdata/builtin_function#L686)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> We should add a test for the `errArgTooBig` case.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822)*

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 1406 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ1yGMH18xlY-grW3xL:-KQ6WeVW1URbUL-VoBAO:b-e041q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1406)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'm not comfortable using switch here because switch suggests the conditions are mutually exclusive and thus that the ordering of tests is indeterminate. Here the conditions have to be evaluated in sequence.
</details>
A switch statement still gives you that property, but ok.

---

*[sql/parser/builtins.go, line 1436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ18qq-rqz1xgqhDRva-r1-1436:-KQ6XD8jSZ8ZkTlHxvl9:bvpfb1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1436)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Thanks for the tip! I'm using the Sizeof solution now.
</details>
We might want to just export this from `util/decimal` as `decimal.WordSize` and use that here and in the `encoding` package. No strong feelings.

---

*[sql/parser/builtins.go, line 1453 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ2-_LdLg1W6LQ87lOc:-KQ6XXy72woTXuTP60Yz:bdu5b1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1453)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
Much better, thanks,

---

*[sql/parser/builtins.go, line 1420 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ6WzKTOXJA8XVzr-HE:-KQ6WzKTOXJA8XVzr-HF:b7ma8oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/922ceef211cb2e37ac6e43fd9f0d127a9068f762/sql/parser/builtins.go#L1420)):*
> ```Go
> 	v, frac := math.Modf(x * pow)
> 	// The following computation implements unbiased rounding, also called
> 	// bankers' rounding. It ensures that when values that fall exactly
> ```

`that when values that`? The second sentence doesnt quite make sense to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-:-KQ6XbSfSrOyPS1Otv8J:bnfp4nl)*

**Comment by knz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 1436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ18qq-rqz1xgqhDRva-r1-1436:-KQAfNxvpbO9cAxsjaDW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f069ee71016263ccfcd33bf32104f4ae1f193014/sql/parser/builtins.go#L1436)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> We might want to just export this from `util/decimal` as `decimal.WordSize` and use that here and in the `encoding` package. No strong feelings.
</details>
Done.

---

*[sql/parser/builtins.go, line 1420 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822#-KQ6WzKTOXJA8XVzr-HE:-KQAekR4DEiPHJgm8UyF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/922ceef211cb2e37ac6e43fd9f0d127a9068f762/sql/parser/builtins.go#L1420)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `that when values that`? The second sentence doesnt quite make sense to me.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8822)*
